pkg/providers/snow/objects.go

```diff
@@ -156,3 +166,28 @@ func recreateKubeadmConfigTemplateNeeded(new, old *bootstrapv1.KubeadmConfigTemp
 	}
 	return !equality.Semantic.DeepDerivative(new.Spec, old.Spec)
 }
+
+// credentialsSecret generates the credentials secret(s) used for provisioning a snow cluster.
+// - eks-a credentials secret: user managed secret referred from snowdatacenterconfig identityRef
+// - snow credentials secret: eks-a creates, updates and deletes in eksa-system namespace. this secret is fully managed by eks-a. User shall treat it as a "read-only" object
+func credentialsSecrets(ctx context.Context, clusterSpec *cluster.Spec) ([]kubernetes.Object, error) {
+	// TODO: controller create workload cluster, should we throw an error? or print a warning?
+	if clusterSpec.SnowCredentialsSecret == nil {
+		return []kubernetes.Object{}, nil
+	}
+
+	// we reconcile the snow credentials secret to be in sync with the eks-a credentials secret user manages.
+	// notice for cli upgrade, we handle the eks-a credentials secret update in a separate step - under provider.UpdateSecrets
+	// which runs before the actual cluster upgrade.
+	// for controller secret, the user is responsible for making sure the eks-a credentials secret is created and up to date.
+	credsB64, ok := clusterSpec.SnowCredentialsSecret.Data["credentials"]
+	if !ok {
+		return nil, fmt.Errorf("unable to retrieve credentials from secret [%s]", clusterSpec.SnowCredentialsSecret.GetName())
+	}
+	certsB64, ok := clusterSpec.SnowCredentialsSecret.Data["ca-bundle"]
+	if !ok {
+		return nil, fmt.Errorf("unable to retrieve ca-bundle from secret [%s]", clusterSpec.SnowCredentialsSecret.GetName())
+	}
+
+	return []kubernetes.Object{clusterSpec.SnowCredentialsSecret, SnowCredentialsSecret(clusterSpec, string(credsB64), string(certsB64))}, nil
```

Comment:
why do we return `clusterSpec.SnowCredentialsSecret`?
I would expect that to already exist

Reply:
this might or might not exist -- for cli create, we need to create this object and push to the cluster. Also during CLI upgrade, if user changes ENVs, we need to update this eksa obj as well.

Reply:
Added a different comment talking about this same thing and proposing a different solution. Closing this one.